js/components/util/promises.ts

```diff
@@ -11,6 +11,26 @@ export function noAwait<F extends (...args: unknown[]) => Promise<unknown>>(f: F
     return f as NoAwait<F>;
 }
 
+/** Returns true if the spawned task should stop doing work. */
+export type Interrupted = () => boolean;
+
+/** Tells the spawned task to stop doing work. */
+export type Interrupt = () => void;
+
+/**
+ * Calls the given `Promise`-returning function, `f`, and does not await
+ * the result. The function `f` is passed an {@link Interrupted} function
+ * that returns `true` if it should stop doing work.  `spawn` returns an
+ * {@link Interrupt} function that, when called, causes the `Interrupted`
+ * function to return `true`. This can be used in `useEffect` calls as the
+ * cleanup function.
+ */
+export function spawn(f: (interrupted: Interrupted) => Promise<unknown>): Interrupt {
+    let interrupted = false;
+    noAwait(f)(() => interrupted);
+    return () => { interrupted = true; };
```

Comment:
Could we borrow a page from [AbortSignal](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) and include `reason`? (I poked around and tried to determine if repurposing `AbortSignal` outside it's current intended purpose was advisable and it didn't seem encouraging)

Reply:
Thanks for pointing me at `AbortSignal` and `AbortController`—I wasn't aware of them. They certainly provide some interesting possibilities.

From what I understand, passing a reason to `abort(...)` causes that reason to be thrown when used in conjunction with a long-running DOM operation like `fetch`. `AbortSignal` also has `throwIfAborted` which throws the reason if the signal is aborted. This seems reasonable as the goal is to stop an operation abruptly and the only real way to signal that is with an exception. Also, if I read things right, the only way to create an `AbortSignal` is by creating an `AbortController`; it has no public constructor.

For this implementation of `spawn`, I was trying to keep things as simple and lightweight as possible and just trying to "gently" signal that the async block should stop. I did this based on the discussions I read about why `useEffect` doesn't take async functions, namely that if the function is called several times, the invocations could step on each other if they complete out of order.

But let's push the `AbortController` idea and see what happens.

Let's imagine that we have something like:

```typescript
    useEffect(() => {
        for (const drive of [1, 2] as DriveNumber[]) {
            spawn(async () => {
                const file = await downloadHttpFile(hashPart, onProgress);
                await smartStorageBroker.loadFile(drive, file);
                onProgress(0, 0);
             });
        }
    }, [hash, onProgress, ready, storageDevices]);
```

The goal here would be to allow the `useEffect` cleanup task to also abort the download. Classically, it would look like:

```typescript
    useEffect(() => {
        const abortController = new AbortController();
        for (const drive of [1, 2] as DriveNumber[]) {
            spawn(async () => {
                try {
                    const file = await downloadHttpFile(hashPart, onProgress, abortController.signal);
                    await smartStorageBroker.loadFile(drive, file, abortController.signal);
                } finally {
                    onProgress(0,0);
                }
             });
        }
        return () => { abortController.abort(); };
    }, [hash, onProgress, ready, storageDevices]);
```

This would work to abort the download and stop loading the file (assuming that the `loadFile` method actually checks the signal state). It also assumes that methods that terminate abruptly will throw.

If we push the abort controller into the `spawn()`:

```typescript
    useEffect((abortSignal) => {
        const controllers: AbortController[] = [];
        for (const drive of [1, 2] as DriveNumber[]) {
            const controller = spawn(async () => {
                try {
                    const file = await downloadHttpFile(hashPart, onProgress, abortSignal);
                    await smartStorageBroker.loadFile(drive, file, abortSignal);
                } finally {
                    onProgress(0,0);
                }
            });
           controllers.push(controller);
        }
        return () => { controllers.forEach(c => c.abort()); };
    }, [hash, onProgress, ready, storageDevices]);
```

It's a bit messier because each nested `spawn` has its own controller. This is also true with just the plain interrupt version, so it's more a problem of having `spawn` do too much. That's not to say that we couldn't do both as nothing forces the caller to use the controller or signal. Also, it appears that chaining using `addEventListener` works, too.

I've made these changes and added a few tests for illustration. Somewhat ironically, I had objects very much like `AbortController` and `AbortSignal` in the first implementation and removed them for the simplicity of functions. :-)

Also, we should watch https://github.com/tc39/proposal-cancellation.